### PR TITLE
Refactor: Add shortcut manager to replace shortcut callbacks in flux

### PR DESCRIPTION
**Pull request by meiliang1123**
- remove some old shortcuts

### CODE REVIEW

File: src/app/containers/App.jsx [hunk 1]
```diff
@@ -83,6 +82,36 @@ class App extends PureComponent {
         recoveringProject: false
     };
 
+
+    /**
+     * responder.active probably decided by component status,
+     * so we use get property to get active status.
+     * but get property function is binded to responder,
+     * we use closure function or define in constructor to solve this problem.
+     *  */
```
Comment (parachvte) [inline src/app/containers/App.jsx:91]: **/

File: src/app/flux/printing/index.js [hunk 2]
```diff
@@ -948,6 +952,11 @@ export const actions = {
         dispatch(actions.render());
     },
 
+    offsetGcodeLayers: (offset) => (dispatch, getState) => {
```
Comment (parachvte) [inline src/app/flux/printing/index.js:955]: If only `offsetGcodeLayers(1)` and `offsetGcodeLayers(-1)` are used, I would prefer ~~`nextLayer()` and `previousLayer()`.~~ `upperLayer()` and `lowerLayer()`, only 2 functions exposed.

Comment (parachvte) [reply src/app/flux/printing/index.js:955]: If use `offsetGcodeLayers()`, better it being documented.

File: src/app/lib/shortcut/manager.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    for (const comboKey of comboKeys) {
+        keyMap[comboKey.toLocaleLowerCase()] = true;
+    }
+    return Object.keys(keyMap);
+}
+
+class CShortcutManger {
+    _responders = [];
+
+    _comboKeyCallbackMap={};
```
Comment (parachvte) [inline src/app/lib/shortcut/manager.js:16]: make sure to run ESLint before you commit.

File: src/app/lib/shortcut/manager.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+     *  title: string(option),
+     *  priority: number,
+     *  active: boolean,
+     *  shortcuts:{
+     *      [actionName]:function || {keys,callback:function}
+     *  }
+     * } responder : readonly
+     */
+    register(responder) {
+        if (!(responder.priority in Object.values(priorities))) {
```
Comment (parachvte) [inline src/app/lib/shortcut/manager.js:30]: - Check if responder is null
- Check all properties required, `title`? `shortcuts`? and check type of shortcut function

File: src/app/lib/shortcut/manager.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            }
+            comboKeys = parseCombokeys(comboKeys);
+            this._bind(comboKeys, responder, actionName);
+        }
+    }
+    // TODO: unregister
+    // responder should be unregistered when parent component destroyed
+
+    printList() {
+        this._responders.sort((a, b) => a.priority - b.priority);
```
Comment (parachvte) [inline src/app/lib/shortcut/manager.js:63]: Add a comment above to indicate responders are sorted by priority (low to high or high to low)

File: src/app/lib/shortcut/manager.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        const titles = [];
+        for (const responder of this._responders) {
+            if (titles.indexOf(responder.title) === -1) {
```
Comment (parachvte) [inline src/app/lib/shortcut/manager.js:66]: - So title is the unique key for responders?
- Maybe rename `responder` to `shortcutHandler`, it's way more easier to understand and remember.

**Review by parachvte — CHANGES_REQUESTED** [review]
- Only add withRouter once, and convert `ownProps.location.pathname` to the page name (constant) in redux state. Watch redux page state in component if needed.
- Use method below to define responder. Here we want to avoid using enclosure, because it's NOT friendly to semantic analysis (ESLint checking).

```JS
class Visualizer extends PureComponent {
    shortcutResponder = {
        title: this.constructor.name,
        active: this.active,
    };

    get active() {
        return this.isActive;
    }
}
```

File: src/app/lib/shortcut/manager.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                    comboKeys = shortcut.keys;
+                } else if (actionName in actionKeys) {
+                    comboKeys = actionKeys[actionName];
+                }
+                if (!comboKeys) {
+                    continue;
+                }
+                if (typeof actionName === 'symbol') actionName = actionName.toString();
+
+                console.log(`${responder.title}-P${responder.priority} ${actionName} [${comboKeys.join(' , ')}]`);
```
Comment (parachvte) [inline src/app/lib/shortcut/manager.js:87]: remove log

File: src/app/lib/shortcut/manager.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            }
+        }
+    }
+
+    _bind(comboKeys, responder, actionName) {
+        for (const comboKey of comboKeys) {
+            let mapItem = this._comboKeyCallbackMap[comboKey];
```
Comment (parachvte) [inline src/app/lib/shortcut/manager.js:94]:
```JS
if (!this._comboKeyCallbackMap[comboKey]) {
    this._comboKeyCallbackMap[comboKey] = [];
}
const mapItem = this._comboKeyCallbackMap[comboKey];
```

File: src/app/flux/printing/index.js (hunk 2 above)
Comment (meiliang1123) [reply src/app/flux/printing/index.js:955]: use `offsetGcodeLayers()`  because offset should be set larger if shortcut pressed rapidly.

**Comment by meiliang1123** [comment]
> * Only add withRouter once, and convert `ownProps.location.pathname` to the page name (constant) in redux state. Watch redux page state in component if needed.
> * Use method below to define responder. Here we want to avoid using enclosure, because it's NOT friendly to semantic analysis (ESLint checking).
> 
> ```js
> class Visualizer extends PureComponent {
>     shortcutResponder = {
>         title: this.constructor.name,
>         active: this.active,
>     };
> 
>     get active() {
>         return this.isActive;
>     }
> }
> ```

good code! thanks!

File: src/app/lib/shortcut/manager.js (hunk 6 above)
Comment (meiliang1123) [reply src/app/lib/shortcut/manager.js]: title is just for printing, not the unique key for shortcut manger

**Pull request merged**